File: src/main/java/com/jozufozu/exnihiloomnia/common/registries/WeightedDrop.java
package com.jozufozu.exnihiloomnia.common.registries;

import com.google.gson.JsonObject;
import com.google.gson.JsonParseException;
import com.jozufozu.exnihiloomnia.common.lib.LibRegistries;
import com.jozufozu.exnihiloomnia.common.util.IRewardProcessor;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.util.JsonUtils;

import java.util.Random;

public class WeightedDrop
{
    private ItemStack drop;
    private float chance;
    
    private String category;
    
    public WeightedDrop() {}
    
    public WeightedDrop(ItemStack drop, float chance)
    {
        this.drop = drop;
        this.chance = chance;
    }
    
    public ItemStack getDrop()
    {
        return drop;
    }
    
    public float getChance()
    {
        return chance;
    }
    
    public String getType()
    {
        return category;
    }
    
    public ItemStack roll(EntityPlayer user, ItemStack activeStack, Random random)
    {
        float chance = this.chance;
        if (activeStack.getItem() instanceof IRewardProcessor)
            chance *= ((IRewardProcessor) activeStack.getItem()).getEffectivenessForType(this.category);
        
        return random.nextFloat() < chance ? drop.copy() : ItemStack.EMPTY;
    }
    
    public static WeightedDrop deserialize(JsonObject object) throws JsonParseException
    {
        WeightedDrop out = new WeightedDrop();
        
        out.drop = JsonHelper.deserializeItem(object, true);
        out.chance = JsonUtils.getFloat(object, LibRegistries.CHANCE, 1.0f);
        out.category = JsonUtils.getString(object, LibRegistries.DROP_CATEGORY, "");
        
        return out;
    }
}

File: src/main/kotlin/com/jozufozu/exnihiloomnia/common/registries/WeightedDrop.kt
package com.jozufozu.exnihiloomnia.common.registries

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParseException
import com.google.gson.JsonSyntaxException
import com.jozufozu.exnihiloomnia.common.lib.LibRegistries
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack
import net.minecraft.util.JSONUtils
import net.minecraft.util.ResourceLocation
import net.minecraftforge.registries.ForgeRegistries
import java.util.*

class WeightedDrop(
        val drop: ItemStack,
        val chance: Float = 0.0f,
        val type: String = ""
) {

    fun roll(user: PlayerEntity?, multipliers: Map<String, Float>, random: Random): ItemStack {
        var chance = this.chance
        user?.let { chance += it.luck }
        chance *= multipliers.getOrDefault(type, 1f)

        return if (random.nextFloat() < chance) drop.copy() else ItemStack.EMPTY
    }

    companion object {

        @Throws(JsonParseException::class)
        fun deserialize(drop: JsonElement): WeightedDrop {
            return when {
                drop is JsonObject -> {
                    val item = JsonHelper.deserializeItem(drop, true)
                    val chance = JSONUtils.getFloat(drop, LibRegistries.CHANCE, 1.0f)
                    val type = JSONUtils.getString(drop, LibRegistries.DROP_CATEGORY, "")

                    WeightedDrop(item, chance, type)
                }
                drop.isJsonPrimitive && drop.asJsonPrimitive.isString -> {
                    val itemName = drop.asJsonPrimitive.asString
                    val itemMaybe = ForgeRegistries.ITEMS.getValue(ResourceLocation(itemName))
                            ?: throw JsonParseException("unknown item '$itemName'")

                    WeightedDrop(ItemStack(itemMaybe), 1f)
                }
                else -> throw JsonSyntaxException("reward must be either a string or an object")
            }
        }
    }
}
